WireAPI/Sources/WireAPI/Network/APIService/APIService.swift

```diff
@@ -0,0 +1,131 @@
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+// sourcery: AutoMockable
+/// A service for network communication to a specific backend.
+///
+/// An api service allows api clients to communicate to a target backend.
+/// It may provide additional functionality, such providing authentication
+/// headers if needed.
+public protocol APIServiceProtocol {
+
+    /// Execute a request to the backend.
+    ///
+    /// - Parameters:
+    ///   - request: A url request.
+    ///   - requiringAccessToken: Whether the request requires an access token.
+    ///
+    /// - Returns: The response to the request.
+
+    func executeRequest(
+        _ request: URLRequest,
+        requiringAccessToken: Bool
+    ) async throws -> (Data, HTTPURLResponse)
+
+}
+
+/// A service for network communication to a specific backend.
+///
+/// An api service allows api clients to communicate to a target backend.
+/// It may provide additional functionality, such providing authentication
+/// headers if needed.
+
+public final class APIService: APIServiceProtocol {
+
+    private let backendURL: URL
+    private let backendWebSocketURL: URL
+    private let authenticationStorage: any AuthenticationStorage
+    private let urlSession: any URLSessionProtocol
+
+    /// Create a new `APIService`.
+    ///
+    /// - Parameters:
+    ///   - backendURL: The url of the target backend.
+    ///   - backendWebSocketURL: The web socket url of the target backend.
+    ///   - authenticationStorage: The storage for authentication objects.
+    ///   - minTLSVersion: The minimum supported TLS version.
+
+    public convenience init(
+        backendURL: URL,
+        backendWebSocketURL: URL,
+        authenticationStorage: any AuthenticationStorage,
+        minTLSVersion: TLSVersion
+    ) {
+        let configFactory = URLSessionConfigurationFactory(minTLSVersion: minTLSVersion)
+        let configuration = configFactory.makeRESTAPISessionConfiguration()
+        let urlSession = URLSession(configuration: configuration)
+
+        self.init(
+            backendURL: backendURL,
+            backendWebSocketURL: backendWebSocketURL,
+            authenticationStorage: authenticationStorage,
+            urlSession: urlSession
+        )
+    }
+
+    init(
+        backendURL: URL,
+        backendWebSocketURL: URL,
+        authenticationStorage: any AuthenticationStorage,
+        urlSession: any URLSessionProtocol
+    ) {
+        self.backendURL = backendURL
+        self.backendWebSocketURL = backendWebSocketURL
+        self.authenticationStorage = authenticationStorage
+        self.urlSession = urlSession
+    }
+
+    /// Execute a request to the backend.
+    ///
+    /// - Parameters:
+    ///   - request: A url request.
+    ///   - requiringAccessToken: Whether the request requires an access token.
+    ///
+    /// - Returns: The response to the request.
+
+    public func executeRequest(
```

Comment:
What do we do with the responses of our requests via `APIService`? Do they tend to always be JSON data? I'm wondering if it is worth incorporating JSON parsing here or in a convenience method. I quite like the approach of https://github.com/objcio/tiny-networking (which I'm not sure is production ready) where defining a new endpoint takes very little code. I think this kind generic approach may require less code and testing then the BackendInfoAPIBuilder pattern we are using. But it does give up some flexibility. 

Anyway I'm just thinking out loud and this is not a request for change. Please take a look at the link above. And feel free to close this comment.

Reply:
In most cases they will be json payloads, but not always. I feel like if we include JSON parsing here, then the api service would then become aware of which responses are considered successes and which are not, which to mean is what the api clients are for.

I like the idea of representing endpoints as objects, but I'm not sure how it would handle API versioning. I suppose the most obvious approach would be to switch over an api version inside the endpoint object, but that resembles our existing code that uses a lot of switches to change behavior depending on the version. 

In the end it's a question about how you want to slice the code up, my personal preference is to keep the implementation of a version contained a grouped together. But I'm open to considering alternatives and would be happy to discuss it further on a call.

Reply:
👍 Thanks for the detailed response. Good point about API versioning - this is something I have never had to consider.

I think your solution is good and it is probably not worth overthinking this until we see how things work in practice. It is likely that we need a lot of flexibility which your solution provides.